### ISSUE TITLE
`needless_return`  has false negative pattern

### ISSUE THREAD

**Issue by kyoto7250** [rust-lang/rust-clippy#8978]
### Summary

There is a pattern that does not work even though needless_return is enabled.

[playground is here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9f84d1d415ac46d0e0fdb123b4a46462).

### Lint Name

needless_return

### Reproducer

I tried this code:

```rust
#![warn(clippy::needless_return)]
const WORDS: [&[&str]; 2] = [&["foo"], &["bar"]];

fn main() {
    println!("{:?}", issue());
}

fn issue() -> bool {
    return WORDS.iter().any(|&word| word == ["foo"]);
}
```

I expected to see this happen: There should be a suggestion as above follows.

```rust
fn issue() -> bool {
    WORDS.iter().any(|&word| word == ["foo"])
}
```

Instead, this happened: No happen.

### Version

```text
rustc 1.61 (stable)
clippy 1.63
```

**Comment by torokati44** [comment]
A bit more simplified reproducer:
```rust
#![warn(clippy::needless_return)]
fn main() {
    println!("{:?}", issue());
}

fn issue() -> bool {
    return [].iter().any(|_: &i32| false);
}
```

Referencing: https://github.com/ruffle-rs/ruffle/pull/16906#discussion_r1657888561

**Comment by samueltardieu** [comment]
This seems to be fixed. Can we close it?

**Comment by kyoto7250** [comment]
@samueltardieu 
Yes, I checked this issue is resolved in latest version. 
I'll close it. Thank you 👍🏽
